### ISSUE TITLE
CS8602 false positive after Debug.Assert() on .NET Standard 2.0

### ISSUE THREAD

**Issue by f-i-x-7** [dotnet/roslyn#57438]
Sorry for possible duplicate, there are 400+ open issues and 1500+ closed ones related to NRT feature.

**Version Used**: 
Visual studio Version 17.0.0 Preview 6.0

dotnet --list-sdks output:
```
3.1.414 [C:\Program Files\dotnet\sdk]
5.0.208 [C:\Program Files\dotnet\sdk]
5.0.303 [C:\Program Files\dotnet\sdk]
5.0.402 [C:\Program Files\dotnet\sdk]
6.0.100-preview.3.21202.5 [C:\Program Files\dotnet\sdk]
6.0.100-rc.2.21505.57 [C:\Program Files\dotnet\sdk]
```

**Steps to Reproduce**:

Compile the following code with NRT feature enabled and target frameworks netstandard2.0, netcoreapp3.1, net5.0, net6.0:

```c#
    public sealed class Foo<T>
    {
        private readonly Func<int, T>? _func1;
        private readonly Func<T>? _func2;

        public Foo(Func<int, T>? func)
        {
            _func1 = func ?? throw new ArgumentNullException(nameof(func));
        }

        public Foo(Func<T>? func)
        {
            _func2 = func ?? throw new ArgumentNullException(nameof(func));
        }

        public T Bar()
        {
            if (_func1 is not null)
            {
                return _func1(42);
            }

            Debug.Assert(_func2 is not null); // Should eliminate warning CS8602 Dereference of a possibly null reference
            return _func2(); // But actually warning is shown for .NET Standard 2.0
        }
    }
```

See also repro repository https://github.com/f-i-x-7/IssuesForDotnetRoslyn

**Expected Behavior**:
There is no CS8602 warning for all mentioned TFMs because latest SDK/compiler version is used and it should recognize that after `Debug.Assert()` the field `_func2` should not be treated as potentially null reference.

**Actual Behavior**:
CS8602 warning is shown at line `return _func2();` for netstandard2.0 TFM. No warnings for netcoreapp3.1/net5.0/net6.0 TFMs.
